File: backend/app/schemas/circuit.py
from __future__ import annotations
from pydantic import BaseModel, Field


# ── Ohm's Law ──────────────────────────────────────────────────────────────────

class OhmsRequest(BaseModel):
    voltage: float | None = Field(None, ge=0.001, le=10000)
    current: float | None = Field(None, ge=0.000001, le=1000)
    resistance: float | None = Field(None, ge=0.001, le=10_000_000)
    time: float = Field(1.0, ge=0.001, le=3600, description="Duration for energy calc (s)")

    model_config = {"json_schema_extra": {"example": {"voltage": 12.0, "resistance": 100.0}}}


class OhmsResultResponse(BaseModel):
    voltage: float
    current: float
    resistance: float
    power: float
    energy: float
    v_drop: float
    iv_curve: list[list[float]]


class OhmsResponse(BaseModel):
    experiment_id: str
    parameters: dict
    results: OhmsResultResponse


# ── Series / Parallel ─────────────────────────────────────────────────────────

class SeriesParallelRequest(BaseModel):
    resistances: list[float] = Field(..., min_length=1, max_length=6)
    supply_voltage: float = Field(..., ge=0.001, le=10000)
    topology: str = Field("series", pattern="^(series|parallel|series_parallel)$")


class ResistorInfoResponse(BaseModel):
    id: int
    resistance: float
    voltage_drop: float
    current: float
    power: float


class SeriesParallelResultResponse(BaseModel):
    topology: str
    supply_voltage: float
    equivalent_resistance: float
    total_current: float
    total_power: float
    resistors: list[ResistorInfoResponse]
    voltage_drops: list[dict]
    power_dist: list[dict]


class SeriesParallelResponse(BaseModel):
    experiment_id: str
    parameters: dict
    results: SeriesParallelResultResponse


# ── RC Circuit ────────────────────────────────────────────────────────────────

class RCRequest(BaseModel):
    resistance: float = Field(..., ge=1, le=10_000_000, description="Resistance (Ω)")
    capacitance: float = Field(..., ge=1e-9, le=1.0, description="Capacitance (F)")
    supply_voltage: float = Field(..., ge=0.001, le=10000, description="Supply voltage (V)")
    initial_voltage: float = Field(0.0, ge=0.0, le=10000, description="Initial cap voltage (V)")
    mode: str = Field("charging", pattern="^(charging|discharging)$")


class RCResultResponse(BaseModel):
    resistance: float
    capacitance: float
    supply_voltage: float
    time_constant: float
    mode: str
    voltage_at_tau: list[float]
    time_series: list[list[float]]
    charge_at_5tau: float
    energy_stored: float


class RCResponse(BaseModel):
    experiment_id: str
    parameters: dict
    results: RCResultResponse

File: frontend/src/types/circuit.ts
// ── Ohm's Law ──────────────────────────────────────────────────────────────────

export interface OhmsRequest {
  voltage?: number;
  current?: number;
  resistance?: number;
  time?: number;
}

export interface OhmsResult {
  voltage: number;
  current: number;
  resistance: number;
  power: number;
  energy: number;
  v_drop: number;
  iv_curve: [number, number][];
}

export interface OhmsResponse {
  experiment_id: string;
  parameters: Record<string, number | null>;
  results: OhmsResult;
}

// ── Series / Parallel ─────────────────────────────────────────────────────────

export type Topology = "series" | "parallel" | "series_parallel";

export interface SeriesParallelRequest {
  resistances: number[];
  supply_voltage: number;
  topology: Topology;
}

export interface ResistorInfo {
  id: number;
  resistance: number;
  voltage_drop: number;
  current: number;
  power: number;
}

export interface SeriesParallelResult {
  topology: Topology;
  supply_voltage: number;
  equivalent_resistance: number;
  total_current: number;
  total_power: number;
  resistors: ResistorInfo[];
  voltage_drops: { name: string; value: number }[];
  power_dist: { name: string; value: number }[];
}

export interface SeriesParallelResponse {
  experiment_id: string;
  parameters: Record<string, unknown>;
  results: SeriesParallelResult;
}

// ── RC Circuit ────────────────────────────────────────────────────────────────

export type RCMode = "charging" | "discharging";

export interface RCRequest {
  resistance: number;
  capacitance: number;
  supply_voltage: number;
  initial_voltage?: number;
  mode: RCMode;
}

export interface RCResult {
  resistance: number;
  capacitance: number;
  supply_voltage: number;
  time_constant: number;
  mode: RCMode;
  voltage_at_tau: number[];
  time_series: [number, number, number][];  // [t, Vc, Ic]
  charge_at_5tau: number;
  energy_stored: number;
}

export interface RCResponse {
  experiment_id: string;
  parameters: Record<string, unknown>;
  results: RCResult;
}
